Csvup 8 feature keys

dklein1211 opened the pull request.

dklein1211 committed:
FEATURE: Added the ability to create users without devices.

TESTING: Console logs left in for reviewer testing.
-Single User with a device: PASSED
-Single User without a device: PASSED
-Multi-User with devices: PASSED
-Multi-User with devices and only users: PASSED
-Multi-User with devices with missing information: PASSED

dklein1211 committed:
FEATURE: When adding a user with a device or not the user will be added to the company directory automatically.

TESTING: Logs removed
-Single user with a device - PASSED
-Single user without a device - PASSED
-Multiple users with devices - PASSED
-Multiple users with no devices - PASSED
-Multiple users with and without devices - PASSED

dklein1211 committed:
FEATURE: Added the ability to create an optional softphone for a user with a yes or no option and cleaned up some code to work better for adding an optional device.

TESTED:
-Extensively tested all scenarios on/off - PASSED
-Created No device but softphone - PASSED

dklein1211 committed:
FEATURE: Added provisioner device validation. This feature will match the devices in the CSV with the Audian provisioner API call.

TEST:
-Build user with correct brand/family/model - PASSED
-Do not build user with incorrect brand - PASSED
-Do not build user with incorrect family - PASSED
-Do not build user with incorrect model - PASSED

dklein1211 committed:
Feature Added: Ability to add feature keys to new users with devices added to the account.

dklein1211 committed:
FIX: Updated api call that was incorrect.

AnthonyPerish commented:
Feature keys bug: If I set a feature key, then change the feature key to none, an error is thrown when trying to save.

![Screen Shot 2019-09-06 at 10 37 38 AM](https://user-images.githubusercontent.com/21110880/64450331-c2009280-d096-11e9-8b33-61dba0fff991.png)

![Screen Shot 2019-09-06 at 10 38 36 AM](https://user-images.githubusercontent.com/21110880/64450334-c4fb8300-d096-11e9-9349-85e7c2faf185.png)

![Screen Shot 2019-09-06 at 10 38 57 AM](https://user-images.githubusercontent.com/21110880/64450357-d3499f00-d096-11e9-8103-cf026320f4c7.png)

AnthonyPerish commented:
Feature Keys Bug: Device update API is being called twice for each device which often times results in a datastore conflict due to updating the device twice in a row too quick.  This seems to happen every time for me when I have multiple users/devices.  Screenshots attached are from clicking proceed once. 

![Screen Shot 2019-09-06 at 10 55 54 AM](https://user-images.githubusercontent.com/21110880/64450505-23286600-d097-11e9-8888-ef235de546d6.png)

![Screen Shot 2019-09-06 at 10 56 14 AM](https://user-images.githubusercontent.com/21110880/64450523-2885b080-d097-11e9-9977-e4f64d400ac2.png)

AnthonyPerish commented:
On initial screen upon loading app the "upload a CSV File with the following columns" text needs to be updated to include the new column softphone(yes|no).

AnthonyPerish commented:
When trying to create a user with no device, the following error is thrown about the brand, plus the proceed button is disabled.  In the instance when nothing is specified for a device this should not happen.

![Screen Shot 2019-09-06 at 11 25 30 AM](https://user-images.githubusercontent.com/21110880/64451512-3b00e980-d099-11e9-8498-0ce09952d7fd.png)

AnthonyPerish commented on app.js:
```diff
@@ -486,7 +475,250 @@ define(function(require) {
+				//Verify if the targeted device (deviceInfo) has been added to the device list (deviceModels).
+				//This is used narrow down the API calls to the provisioner to get the max feature keys that device can have.
+				_.each(deviceList, function(listDevice) {
+					var found = false;
```
This whole _.each section seems unnecessarily complex.  Seems you could replace it with if (device.brand !== listDevice.brand || device.family !== listDevice.family || device.model !== listDevice.model { deviceList.push(device); }

dklein1211 replied:
This section has been removed in my latest commit and simplified.

dklein1211 committed:
FIXES:
-Resolved resetting feature keys back to none issue.
-Resolved duplicate API calls error.

dklein1211 commented:
> When trying to create a user with no device, the following error is thrown about the brand, plus the proceed button is disabled. In the instance when nothing is specified for a device this should not happen.
> 
> ![Screen Shot 2019-09-06 at 11 25 30 AM](https://user-images.githubusercontent.com/21110880/64451512-3b00e980-d099-11e9-8498-0ce09952d7fd.png)

dklein1211 commented:
> Feature keys bug: If I set a feature key, then change the feature key to none, an error is thrown when trying to save.
> 
> ![Screen Shot 2019-09-06 at 10 37 38 AM](https://user-images.githubusercontent.com/21110880/64450331-c2009280-d096-11e9-8b33-61dba0fff991.png)
> 
> ![Screen Shot 2019-09-06 at 10 38 36 AM](https://user-images.githubusercontent.com/21110880/64450334-c4fb8300-d096-11e9-9349-85e7c2faf185.png)
> 
> ![Screen Shot 2019-09-06 at 10 38 57 AM](https://user-images.githubusercontent.com/21110880/64450357-d3499f00-d096-11e9-8103-cf026320f4c7.png)

This has been resolved in the current commit

dklein1211 committed:
FIX 1: None and blank as options for device brands will result in no device being made and if no devices are being added skip the feature key part.
TEST 1: Created 2 users with a none and a blank brand cell. -  PASSED
TEST 2: Created 3 users with 1 user that has a valid device and the others have none and a blank brand cell. -  PASSED

FIX 2: Valid devices with no MAC address will make a device as a placeholder.
TEST 1: Created 2 users, one has a valid device with mac and the other had a valid device with no mac - PASS

dklein1211 committed:
FIX: Cleaned up console.logs and updated article link.
